File: src/main/java/com/voyanta/service/VehicleService.java
package com.voyanta.service;

import com.voyanta.dto.response.VehicleResponse;
import com.voyanta.entity.Vehicle;
import com.voyanta.repository.FeedbackRepository;
import com.voyanta.repository.VehicleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Transactional
public class VehicleService {
    
    @Autowired
    private VehicleRepository vehicleRepository;
    
    @Autowired
    private FeedbackRepository feedbackRepository;
    
    public List<VehicleResponse> getAllVehicles() {
        return vehicleRepository.findAll().stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public List<VehicleResponse> getAvailableVehicles() {
        return vehicleRepository.findByStatus(Vehicle.VehicleStatus.AVAILABLE).stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public List<VehicleResponse> getVehiclesByType(Vehicle.VehicleType type) {
        return vehicleRepository.findByType(type).stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public List<VehicleResponse> getVehiclesByMake(String make) {
        return vehicleRepository.findByMake(make).stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public List<VehicleResponse> searchVehicles(String searchTerm) {
        return vehicleRepository.findBySearchTerm(searchTerm).stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public List<VehicleResponse> getVehiclesByPriceRange(BigDecimal minRate, BigDecimal maxRate) {
        return vehicleRepository.findByDailyRateBetween(minRate, maxRate).stream()
                .map(this::convertToVehicleResponse)
                .collect(Collectors.toList());
    }
    
    public Optional<Vehicle> getVehicleById(Long id) {
        return vehicleRepository.findById(id);
    }
    
    public VehicleResponse getVehicleResponseById(Long id) {
        Optional<Vehicle> vehicle = vehicleRepository.findById(id);
        return vehicle.map(this::convertToVehicleResponse).orElse(null);
    }
    
    public Vehicle createVehicle(Vehicle vehicle) {
        return vehicleRepository.save(vehicle);
    }
    
    public Vehicle updateVehicle(Vehicle vehicle) {
        return vehicleRepository.save(vehicle);
    }
    
    public void deleteVehicle(Long id) {
        vehicleRepository.deleteById(id);
    }
    
    public void updateVehicleStatus(Long id, Vehicle.VehicleStatus status) {
        Optional<Vehicle> vehicleOpt = vehicleRepository.findById(id);
        if (vehicleOpt.isPresent()) {
            Vehicle vehicle = vehicleOpt.get();
            vehicle.setStatus(status);
            vehicleRepository.save(vehicle);
        }
    }
    
    public List<String> getAllMakes() {
        return vehicleRepository.findDistinctMakes();
    }
    
    public List<String> getModelsByMake(String make) {
        return vehicleRepository.findDistinctModelsByMake(make);
    }
    
    private VehicleResponse convertToVehicleResponse(Vehicle vehicle) {
        VehicleResponse response = new VehicleResponse(vehicle);
        
        // Calculate average rating and total feedbacks
        Double averageRating = feedbackRepository.getAverageRatingByVehicle(vehicle);
        Long totalFeedbacks = feedbackRepository.getFeedbackCountByVehicle(vehicle);
        
        response.setAverageRating(averageRating);
        response.setTotalFeedbacks(totalFeedbacks);
        
        return response;
    }
}

File: frontend/src/services/vehicleService.js
import api from './api';

export const vehicleService = {
  getAllVehicles: async () => {
    const response = await api.get('/public/vehicles');
    return response.data;
  },

  getAvailableVehicles: async () => {
    const response = await api.get('/public/vehicles/available');
    return response.data;
  },

  getVehicleById: async (id) => {
    const response = await api.get(`/public/vehicles/${id}`);
    return response.data;
  },

  getVehiclesByType: async (type) => {
    const response = await api.get(`/public/vehicles/type/${type}`);
    return response.data;
  },

  getVehiclesByMake: async (make) => {
    const response = await api.get(`/public/vehicles/make/${make}`);
    return response.data;
  },

  searchVehicles: async (query) => {
    const response = await api.get(`/public/vehicles/search?q=${query}`);
    return response.data;
  },

  getVehiclesByPriceRange: async (minRate, maxRate) => {
    const response = await api.get(`/public/vehicles/price-range?minRate=${minRate}&maxRate=${maxRate}`);
    return response.data;
  },

  getAllMakes: async () => {
    const response = await api.get('/public/vehicles/makes');
    return response.data;
  },

  getModelsByMake: async (make) => {
    const response = await api.get(`/public/vehicles/models?make=${make}`);
    return response.data;
  },

  // Admin functions
  createVehicle: async (vehicleData) => {
    const response = await api.post('/admin/vehicles', vehicleData);
    return response.data;
  },

  updateVehicle: async (id, vehicleData) => {
    const response = await api.put(`/admin/vehicles/${id}`, vehicleData);
    return response.data;
  },

  deleteVehicle: async (id) => {
    const response = await api.delete(`/admin/vehicles/${id}`);
    return response.data;
  },

  updateVehicleStatus: async (id, status) => {
    const response = await api.put(`/admin/vehicles/${id}/status?status=${status}`);
    return response.data;
  }
};
